### ISSUE TITLE
Encoding in Style: a StyleGAN Encoder for Image-to-Image Translation

### ISSUE THREAD

**Issue by reyllama** [reyllama/paper-reviews#148]
## TL; DR

- Learn a FPN-based encoder that maps input image to the corresponding latent style vectors for the desired output, and feed through the pretrained frozen generator.

## Motivation

- Traditional *Invert First, Edit Later* approach has limitations in that images belonging to the range space of the pretrained generator can be inverted (i.e., there must exist a corresponding latent vector that generates the image once fed through the generator)
- Previous image-to-image translation methods require specific architecture designs and trainings for different tasks, lacking a *generic* framework like *pix2pix*.

## Prior Works

- GAN Inversion: optimization-based (high quality, slow), encoder-based (lower quality, fast)
- Latent Space Manipulation: boundary searching for attributes, principal direction exploration
- Image-to-Image Translation

## Method

![image](https://user-images.githubusercontent.com/46308407/148152456-7edfcec6-f6b7-41d3-b5e6-5cb1031ca17b.png)

- ***Motif***: different style vectors are responsible for different levels of semantics (coarse-medium-fine)
- Extract 3 level feature maps from the given image, and small fully convolutional *map2style* networks transform them to style vectors (total 18 styles for 1024x1024 image, 14 for 256x256)
- In short, pSp operates in *W*+ space for reconstruction of finer details.

![image](https://user-images.githubusercontent.com/46308407/148152941-9623b3ac-2e69-42c3-8772-08540189f443.png)
- Residual modeling (better initialization)

### Objective

- L2 loss
![image](https://user-images.githubusercontent.com/46308407/148152887-98c33ec8-b436-42b4-842a-02caa689b45f.png)

- Perceptual Loss (LPIPS)
![image](https://user-images.githubusercontent.com/46308407/148152894-1825932c-0c72-46c8-850f-c106966499cc.png)

- Latent Regularization (conceptually similar to truncation trick)
![image](https://user-images.githubusercontent.com/46308407/148152903-b33e0b93-875b-431c-9c15-77dc41c564c4.png)

- ID Loss (*face only*)
![image](https://user-images.githubusercontent.com/46308407/148152919-d0985e00-d77d-4272-866a-08b7ab98d98e.png)
- *R* corresponds to pretrained *ArcFace*.

### Inversion (Reconstruction)
![image](https://user-images.githubusercontent.com/46308407/148153036-36e83087-d006-4db0-842d-23d5a911202b.png)
![image](https://user-images.githubusercontent.com/46308407/148153046-68578d19-0071-4a9a-a02c-c819e26fe773.png)

### Qualitative Ablation
![image](https://user-images.githubusercontent.com/46308407/148153241-7a552fda-736e-451e-9670-00355c6243b2.png)

### App: Face Frontalization
![image](https://user-images.githubusercontent.com/46308407/148153262-575ffdd5-131e-4e1a-aabf-362e47b95609.png)
![image](https://user-images.githubusercontent.com/46308407/148153272-32cb359b-2de6-4132-b0bf-15766c99877d.png)

### App: Conditional Image Synthesis
![image](https://user-images.githubusercontent.com/46308407/148153348-3155a95f-687d-4656-b84a-e2cd535d8396.png)
![image](https://user-images.githubusercontent.com/46308407/148153426-6f84a1a9-bee6-4120-8a2c-aad7acb5582f.png)

- Since these tasks are one-to-many in nature, employ style-mixing between learned latents (lower layer) and random latents (higher layer, finer details)

![image](https://user-images.githubusercontent.com/46308407/148153448-49bcfb42-a9b5-4fa6-b0f7-03c1f743beb8.png)

### App: Super-Resolution
![image](https://user-images.githubusercontent.com/46308407/148153619-4822c1ea-46e9-4ba5-b3a7-03b549aa29f7.png)
![image](https://user-images.githubusercontent.com/46308407/148153648-70050594-9ce1-44db-9034-863eefb61cf5.png)
![image](https://user-images.githubusercontent.com/46308407/148153717-f1a432ad-3fe2-4b6d-aa5a-d3edddf34ed9.png)
![image](https://user-images.githubusercontent.com/46308407/148153728-00656f48-3106-4ec5-a65a-adfb8f5b99b6.png)

### App: Inpainting
![image](https://user-images.githubusercontent.com/46308407/148153657-82a5e34d-c868-4fe6-beac-e901d370243b.png)
![image](https://user-images.githubusercontent.com/46308407/148153683-cb797d99-1571-458c-9f81-ca991a7f765d.png)
![image](https://user-images.githubusercontent.com/46308407/148153701-f85d0bc3-e81a-4495-8bdc-9ceec40926de.png)

### App: Interactive Editing
![image](https://user-images.githubusercontent.com/46308407/148153760-25508638-7158-4a21-86c8-58e4945d845c.png)

### Beyond Facial Domain

![image](https://user-images.githubusercontent.com/46308407/148153797-27a4db97-2055-4d76-b3ff-53e6c356a1aa.png)

- ID loss component has been recognized to be the key factor for the success of pSp.
- However, ArcFace-based ID loss is not applicable to domains other than human faces.
- Hence test on animal faces *without the ID loss* for properties like inversion(reconstruction) and applications.

## Comments

- Formulation is very simple and straightforward: since StyleGAN employs style vectors that control different levels of semantic hierarchies, organize the encoder as FPN-variant that extracts style vectors responsible for the corresponding level of semantic details.
- Framing it as generic image-to-image translation method was also quite successful, as one of the core differences of this method compared to other inversion methods is that images out of the range space of pretrained generator can be processed. (i.e., face sketch input for FFHQ-pretrained generator)
- Style mixing is one of the key strengths of StyleGAN, as it enables controlled generation and manipulation. I may have overlooked its significance so far.
- Still, this work has limitations in that it *must* employ a rich and powerful generator, as the output solely depends on the capacity of the generator. 
- In other words, although it can map out-of-domain images to the support set of generator, the *range set* of the framework is kept fixed. Hence, expansions to domains that are innately difficult to train a strong generator is still an important research topic.
